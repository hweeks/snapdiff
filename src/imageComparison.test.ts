import { readFileSync } from 'fs';

const a = require('./fixtures/a.json');
const b = require('./fixtures/b.json');
import { PNG } from 'pngjs';
import { createHash } from 'crypto';

const { imageComparison, generateDiffImages } = require('./imageComparison');

describe('imageComparison Diffing', () => {
  test('two identical JSONs have no differences', () => {
    const output = imageComparison(a, a);
    expect(output.addedImages.length).toBe(0);
    expect(output.addedTests.length).toBe(0);
    expect(output.imageDiffs.length).toBe(0);
    expect(output.removedImages.length).toBe(0);
    expect(output.removedTests.length).toBe(0);
  });
  test('Add a test and an image to an existing test', () => {
    const output = imageComparison(b, a);
    expect(output).toMatchSnapshot({ time: expect.any(Number), });
    expect(output.imageDiffs.length).toBe(0);
    expect(output.removedImages.length).toBe(0);
    expect(output.removedTests.length).toBe(0);
    expect(output.addedImages.length).toBe(1);
    expect(output.addedTests.length).toBe(1);
  });
  test('Remove a test and an image from an existing test', () => {
    const output = imageComparison(a, b);
    expect(output).toMatchSnapshot({ time: expect.any(Number), });
    expect(output.imageDiffs.length).toBe(0);
    expect(output.removedImages.length).toBe(1);
    expect(output.removedTests.length).toBe(1);
    expect(output.addedImages.length).toBe(0);
    expect(output.addedTests.length).toBe(0);
  });
  test('generate PNG image diff file', () => {
    const diff = {
      context: { 
        title: 'example test',
        fullTitle: 'example test',
        file: 'whatever.file'
      },
      expected: 'testimage1',
      actual: 'testimage2',
    }
    generateDiffImages( [ diff ], { localFolder: __dirname + '/fixtures/imagediff' } );
    const png = PNG.sync.read(readFileSync(`${__dirname}/fixtures/imagediff/images/testimage2-testimage1-expected.png`));
    const pngNew = PNG.sync.read(readFileSync(`${__dirname}/fixtures/imagediff/images/testimage2-testimage1.png`));
    const hashExpected = createHash('sha256').update(png.data).digest('hex');
    const hashActual = createHash('sha256').update(pngNew.data).digest('hex');
    expect(hashActual).toBe(hashExpected);
  });
});