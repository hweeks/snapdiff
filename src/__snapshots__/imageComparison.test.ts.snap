// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add a test and an image to an existing test 1`] = `
Object {
  "addedImages": Array [
    Object {
      "context": Object {
        "file": "snapdiff-tests/example.test.js",
        "fullTitle": "basic test super rad",
        "title": "super rad",
      },
      "image": "hijklmn",
    },
  ],
  "addedTests": Array [
    Object {
      "context": Object {
        "file": "snapdiff-tests/example.test.js",
        "fullTitle": "basic test something else",
        "title": "something else",
      },
      "images": Array [
        "opqrstu",
        "vwxyz01",
      ],
    },
  ],
  "imageDiffs": Array [],
  "removedImages": Array [],
  "removedTests": Array [],
  "time": Any<Number>,
}
`;

exports[`Remove a test and an image from an existing test 1`] = `
Object {
  "addedImages": Array [],
  "addedTests": Array [],
  "imageDiffs": Array [],
  "removedImages": Array [
    Object {
      "context": Object {
        "file": "snapdiff-tests/example.test.js",
        "fullTitle": "basic test super rad",
        "title": "super rad",
      },
      "image": "hijklmn",
    },
  ],
  "removedTests": Array [
    Object {
      "context": Object {
        "file": "snapdiff-tests/example.test.js",
        "fullTitle": "basic test something else",
        "title": "something else",
      },
      "images": Array [
        "opqrstu",
        "vwxyz01",
      ],
    },
  ],
  "time": Any<Number>,
}
`;
